external-import/tenable-security-center/app.py

```diff
@@ -0,0 +1,297 @@
+
+        # Update state
+        state.update({"last_run_start_datetime": now_isodatetime})
+        self.helper.set_state(state=state)
+
+        # Update data retrieval start datetime
+        if last_successful_run is not None:
+            self.logger.info(
+                "[CONNECTOR] Connector last run", {"last_run_start_datetime": last_run}
+            )
+            previous_since = str(self.config.tenable_security_center.export_since)
+
+            self.logger.warning(
+                "[CONNECTOR] Connector acquisition SINCE parameter overwritten",
+                {"previous": previous_since, "current": last_successful_run},
+            )
+            self.assets.since_datetime = datetime.fromisoformat(last_successful_run)
+        else:
+            self.logger.info("[CONNECTOR] Connector has never run successfully...")
+            self.assets.since_datetime = (
+                self.config.tenable_security_center.export_since
+            )
+
+        # Initiate a new work
+        self.work_id = self.helper.api.work.initiate_work(
+            self.helper.connect_id, self.helper.connect_name
+        )
+        # Reset cache
+        self._stix_objects_cache.reset()
+        # Reset expectations
+        self._expected_stix_objects = 0
+        self.logger.info(
+            "[CONNECTOR] Running connector...",
+            {"connector_name": self.helper.connect_name},
+        )
+
+    def _finalize_work(self, error_flag: bool) -> None:
+        """Finalize the work process and logs the completion message.
+
+        This method
+            1. Update the connector state depending on the results.
+            2. Marks the work as processed on the OpenCTI platform.
+            3. Logs a message indicating that the connector ran.
+
+        Args:
+            error_flag(bool): The overall processing results (True if OK False otherwise).
+
+        See Also:
+            _initiate_work method
+
+        """
+        state = self.helper.get_state() or {}
+        now_isodatetime = datetime.now(timezone.utc).isoformat()
+        if not error_flag:
+            state.update(
+                {
+                    "last_successful_run_start_datetime": state.get(
+                        "last_run_start_datetime"
+                    ),
+                }
+            )
+        self.helper.set_state(state=state)
+        message = (
+            f"{self.helper.connect_name} connector {'successfully' if not error_flag else ''} run, "
+            f"storing last_run as {now_isodatetime}"
+        )
+        self.logger.info(message)
+
+    def _send_bundle(self, bundle_json: str) -> None:
+        """Send the STIX bundle to the OpenCTI platform and update the total expectation.
+
+        Args:
+            bundle_json(str): The STIX bundle to send.
+
+        Returns:
+           None
+
+        """
+        bundles_sent = self.helper.send_stix2_bundle(
+            bundle=bundle_json,
+            work_id=self.work_id,
+        )
```

Comment:
Could you add here `cleanup_inconsistent_bundle` set to `True`? :)
Since >= 6.3.3, when using send_stix2_bundle, for having a consistent bundle when splitting the data, it will prevent from MISSING_ERROR_REFERENCE when creating a relationship for example

Here are some example of cases

![image](https://github.com/user-attachments/assets/1115e513-7ba7-4310-aa8c-8bd9741d2e83)

```suggestion
        bundles_sent = self.helper.send_stix2_bundle(
            bundle=bundle_json,
            work_id=self.work_id,
            cleanup_inconsistent_bundle=True
        )
```

Reply:
Thanks for your detailed explanation.

The caching mechanism that you are talking about is not in the client-python helper, right ?

As we want to have consistent bundle when the worker consume the RabbitMQ queue, I'm interesting to talk about it in details.

Reply:
@richard-julien supports the drop of deduplication solution